vllm/lora/punica_wrapper/punica_tpu.py

```diff
@@ -316,10 +313,77 @@ def _apply_bias(
 
         return output.view_as(org_output)
 
-    def _update_prefill_metadata(self,
-                                 token_lora_tensor: torch.Tensor) -> None:
+    # This performs the same tensor ops as the base method, except it does them
+    # on the CPU then transfers the results to the TPU
+    def _update_base_metadata(
+        self,
+        mapping: "LoRAMapping",
+        lora_index_to_id: List[Optional[int]],
+        max_loras: int,
+        vocab_size: int,
+        extra_vocab_size: int,
+        long_lora_context: Optional["LongContextLoRAContext"] = None,
+    ):
+        # Make sure we don't accidentally collect outside operations
+        xm.mark_step()
+
+        # Pad the prompt mapping to avoid running into recompiles on the TPU
+        # TODO: Should this happen inside mapping internally? If so how can we
+        # avoid having backend specific LoRAMapping classes?
+        mapping.prompt_mapping = self._pad_prompt_mapping(
+            mapping.prompt_mapping)
+
+        (
+            base_indices,
+            sampler_indices,
+            sampler_indices_padded,
+            embeddings_indices,
+            long_lora_offsets_tensor,
+            indices_len,
+        ) = convert_mapping(
+            mapping,
+            lora_index_to_id,
+            max_loras,
+            vocab_size,
+            extra_vocab_size,
+            "cpu",
+            long_lora_context,
+        )
+        self._token_lora_indices[:base_indices.shape[0]] = base_indices.to(
```

Comment:
are you compiling for base_indices.shape[0] values or is it fixed by lora rank?

Reply:
`base_indices.shape[0]` really just refers to the number of tokens in the batch, so we're not copying extra tokens